context/src/main/java/io/micronaut/runtime/server/GracefulShutdownLifecycle.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.runtime.server;
+
+import io.micronaut.context.LifeCycle;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Stream;
+
+/**
+ * Interface implemented by beans that support graceful shutdown.
+ *
+ * @author Jonas Konrad
+ * @since 4.5.0
+ */
+public interface GracefulShutdownLifecycle {
```

Comment:
not sure lifecycle is the right name here, since this is only graceful shutdown and not startup. Rename to `GracefulShutdownCapable`